### PR TITLE
Ignore replication for noop updates

### LINKED ISSUE #46366: Unexpected request operation type on replica: update
While looking at an unrelated test failure, I've seen the following log line, which indicates a bigger issue with replicating updates. The way it is handled right now is that the replica is failed. I've tried reproducing it, but with no success. A [quick grep across CI logs](https://build-stats.elastic.co/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0),time:(from:now-60d,mode:quick,to:now))&_a=(columns:!(_source),index:f0e86d70-7efd-11e8-bf69-63c8ef516157,interval:auto,query:(language:lucene,query:'log:%22Unexpected%20request%20operation%20type%20on%20replica%22'),sort:!(time,desc))) indicates that this is somewhat of a frequent occurrence though.

```
WARN ][o.e.a.b.TransportShardBulkAction] [integTest-0] [[test_index][0]] failed to perform indices:data/write/bulk[s] on replica [test_index][0], node[Y8t016ljSP6HZZ7bJu8aLg], [R], s[STARTED], a[id=6OT7lxZGQSKASnUk3Az-QQ]
»  org.elasticsearch.transport.RemoteTransportException: [integTest-1][127.0.0.1:37959][indices:data/write/bulk[s][r]]
»  Caused by: java.lang.IllegalStateException: Unexpected request operation type on replica: update
»  	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOpOnReplica(TransportShardBulkAction.java:439) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.bulk.TransportShardBulkAction.performOnReplica(TransportShardBulkAction.java:409) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnReplica(TransportShardBulkAction.java:389) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardOperationOnReplica(TransportShardBulkAction.java:80) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncReplicaAction.onResponse(TransportReplicationAction.java:522) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncReplicaAction.onResponse(TransportReplicationAction.java:496) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShard.lambda$innerAcquireReplicaOperationPermit$27(IndexShard.java:2946) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.ActionListener$3.onResponse(ActionListener.java:113) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:285) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:237) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShard.lambda$acquireReplicaOperationPermit$25(IndexShard.java:2885) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShard.innerAcquireReplicaOperationPermit(IndexShard.java:2985) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.index.shard.IndexShard.acquireReplicaOperationPermit(IndexShard.java:2884) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.support.replication.TransportReplicationAction.acquireReplicaOperationPermit(TransportReplicationAction.java:870) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncReplicaAction.doRun(TransportReplicationAction.java:581) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.action.support.replication.TransportReplicationAction.handleReplicaRequest(TransportReplicationAction.java:481) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:63) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.transport.InboundHandler$RequestHandler.doRun(InboundHandler.java:262) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:769) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»  	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
»  	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
»  	at java.lang.Thread.run(Thread.java:834) [?:?]
```

Observed here: https://gradle-enterprise.elastic.co/s/p2au2dmcrjjm4/console-log/raw

Can you have a look @dnhatn?

### PR DESCRIPTION
**Pull request by dnhatn**
Previously, we ignore replication for noop updates because they do not have sequence numbers. Since #44603, we started assigning sequence numbers to noop updates leading them to be replicated to replicas.

This bug occurs only on 8.0 for it requires two changes: #41065 and #44603.

Closes #46366

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/action/bulk/BulkWithUpdatesIT.java [hunk 1]
```diff
@@ -618,5 +621,31 @@ public void testInvalidIndexNamesCorrectOpType() {
         assertThat(bulkResponse.getItems()[1].getOpType(), is(OpType.UPDATE));
         assertThat(bulkResponse.getItems()[2].getOpType(), is(OpType.DELETE));
     }
+
+    public void testNoopUpdate() {
+        String indexName = "test";
+        createIndex(indexName, Settings.builder().put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1).build());
+        internalCluster().ensureAtLeastNumDataNodes(2);
+        ensureGreen(indexName);
+        IndexResponse doc = index(indexName, "_doc", "1", "{\"user\": \"xyz\"}");
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/action/bulk/BulkWithUpdatesIT.java]: NIT: you could do the document source here and below with the nicer API (I know David recently replaced a bunch of these spots :)):

i.e.:
```java
        IndexResponse doc = index(indexName, "_doc", "1", "user", "xyz");
        assertThat(doc.getShardInfo().getSuccessful(), equalTo(2));
        final BulkResponse bulkResponse = client().prepareBulk()
            .add(new UpdateRequest().index(indexName).id("1").detectNoop(true).doc("user", "xyz"))
```

**Review by original-brownbear — APPROVED** [review]
LGTM, reads much better now as well :)
Thanks @dnhatn !

File: server/src/test/java/org/elasticsearch/action/bulk/BulkWithUpdatesIT.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/action/bulk/BulkWithUpdatesIT.java]: ++ I pushed https://github.com/elastic/elasticsearch/pull/46458/commits/9cd3f93f4e1abfcdb8e76f5e1e8e2d83d5ab7a48.

**Comment by dnhatn** [comment]
The failure is tracked at https://github.com/elastic/elasticsearch/issues/46425.
@elasticmachine run elasticsearch-ci/bwc

**Comment by dnhatn** [comment]
Thanks @original-brownbear :)

**Pull request merged**